src/System.Memory/tests/ReadOnlySpan/NonPortableCast.cs

```diff
@@ -22,14 +22,34 @@ public static void PortableCastUIntToUShort()
         }
 
         [Fact]
-        public static void PortableCastToTypeContainsReferences()
+        public static void NonPortableCastShortToLong()
+        {
+            short[] a = { 0x1234, 0x2345, 0x3456, 0x4567, 0x5678 };
+            ReadOnlySpan<short> span = new ReadOnlySpan<short>(a);
+            ReadOnlySpan<long> asLong = span.NonPortableCast<short, long>();
+
+            Assert.True(Unsafe.AreSame<long>(ref Unsafe.As<short, long>(ref MemoryMarshal.GetReference(span)), ref MemoryMarshal.GetReference(asLong)));
+            asLong.Validate<long>(0x4567345623451234);
```

Comment:
What makes NonPortableCast non-portable?  I'm wondering in what situations we'd expect the resulting asLong to be unusuable?

Reply:
Alignment guarantees. For example, the following is guaranteed to AV on ARM:

```
byte[] span = stackalloc byte[9];
long[] longSpan = span.Slice(1).NonPortableCast<byte, long>();
Interlocked.CompareExchange<long>(ref span[0], 1, 0);
```

Reply:
We may want to rename `NonPortableCast` to `MemoryMarshal.Cast` to make this name look better:

https://github.com/dotnet/corefx/issues/26139#issuecomment-358090527

Reply:
This test is specifically testing that we are doing the rounding correctly when going from smaller T to larger T. It is expected behavior that the last element of the short array is lost.